### PR TITLE
Rewrite and fix hf mfdes auth

### PR DESCRIPTION
**Pull request by bkerler**
- Some CommandOLD and CommandMIX have been ported to CommandNG structure
- Mifare Desfire Algorithm has been rewritten from scratch to make the function more structured
- Sessionkey handing was fixed (was generated wrong for aes)

### CODE REVIEW

**Comment by bkerler** [comment]
Also added further bugfixes, necessary for further hf mfdes improvements

**Comment by bkerler** [comment]
Also added 2k3 des iso support preparation. Both 2k3, 3des and 3k3 algos not yet verified. Currently only des and aes verified to work.

File: client/cmdhfmfdes.c [hunk 1]
```diff
@@ -913,35 +939,37 @@ static int CmdHF14ADesCreateApp(const char *Cmd) {
...
     if (keylen1 < 1) {
         PrintAndLogEx(ERR, "Keysetting2 must have 1 byte length.");
-        return PM3_EINVARG;
+        return PM3_SNONCES;
     }
 
     if (namelen > 16) {
         PrintAndLogEx(ERR, "Name has a max. of 16 bytes length.");
-        return PM3_EINVARG;
+        return PM3_SNONCES;
```
Comment (iceman1001) [inline client/cmdhfmfdes.c:972]: PM3_SNONCES  is a successfull return code,  used when nonce collecting...  Lets keep the PM3_EINVARG

File: client/cmdhfmfdes.c [hunk 2]
```diff
@@ -1029,36 +1054,56 @@ static int CmdHF14ADesFormatPICC(const char *Cmd) {
...
-    uint8_t data[25] = {keylen}; // max length: 1 + 24 (3k3DES)
-    memcpy(data + 1, key, keylen);
-    SendCommandOLD(CMD_HF_DESFIRE_AUTH1, 2, 1, 0, data, keylen + 1);
+    struct {
+        uint8_t mode;
+        uint8_t algo;
+        uint8_t keyno;
+        uint8_t key[24];
+        uint8_t keylen;
+    } PACKED payload;
```
Comment (iceman1001) [inline client/cmdhfmfdes.c:1070]: usually keep the arrays last in the struct.

File: client/cmdhfmfdes.c [hunk 3]
```diff
@@ -1550,28 +1606,42 @@ static int CmdHF14ADesAuth(const char *Cmd) {
...
-    uint8_t data[25] = {keylength}; // max length: 1 + 24 (3k3DES)
-    memcpy(data + 1, key, keylength);
-    SendCommandOLD(CMD_HF_DESFIRE_AUTH1, cmdAuthMode, cmdAuthAlgo, cmdKeyNo, data, keylength + 1);
+    struct {
+        uint8_t mode;
+        uint8_t algo;
+        uint8_t keyno;
+        uint8_t key[24];
+        uint8_t keylen;
+    } PACKED payload;
```
Comment (iceman1001) [inline client/cmdhfmfdes.c:1615]: if you gonna reuse the struct,  I suggest you make a TYPEDEF of it

File: client/cmdhfmfdes.c [hunk 4]
```diff
@@ -36,6 +36,21 @@ uint8_t key_picc_data[16] = { 0x40, 0x41, 0x42, 0x43, 0x44, 0x45, 0x46, 0x47, 0x
...
+    MFDES_AUTH_PICC = 4
+} mifare_des_authmode_t;
+
+typedef enum {
+    MFDES_ALGO_DES = 1,
+    MFDES_ALGO_3DES = 2,
+    MFDES_ALGO_2K3DES = 3,
+    MFDES_ALGO_3K3DES = 4,
+    MFDES_ALGO_AES = 5
+} mifare_des_authalgo_t;
```
Comment (iceman1001) [inline client/cmdhfmfdes.c:52]: if these enums are using on deviceside aswell,  you can add them to a suitable file inside folder  include/

**Comment by bkerler** [comment]
Thanks for the hints :) Fixed them accordingly (and some more stuff)

**Comment by iceman1001** [comment]
SuperB!

**Pull request merged**